feat(MMENG-4247): Add maven-zip-build pipeline

ligangty opened the pull request:
This new build pipeline will do:
* Accept the cachi2 lockfiles for generic-fetcher as source code in git repository
* Prefetch the artifacts in the cachi2 lockfiles in prefetch task
* Build the OCI-artifact of maven zip bundle which is generated by the artifacts fetched through cachi2 in build-maven-zip task

It only uses these four tasks:
* init 0.2
* git-clone 0.1
* prefetch-dependencies 0.1
* build-maven-zip 0.1

The ta version will use the oci-ta versions of tasks instead of above.

MartinBasti commented on pipelines/maven-zip-build-ta/README.md:
```diff
@@ -0,0 +1,148 @@
+# "maven-zip-build-ta pipeline"
```
should be named with suffix `-oci-ta`

ligangty replied:
Fixed in new commit.

chmeliik commented on pipelines/maven-zip-build-oci-ta/kustomization.yaml:
```diff
@@ -0,0 +1,9 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+- ../template-build
```
Maybe using `../maven-zip-build` as the base would be easier?

ligangty replied:
Not sure. Let me have a check.

ligangty replied:
Used ../maven-zip-build in new commit.

chmeliik commented on pipelines/maven-zip-build/patch.yaml:
```diff
@@ -0,0 +1,125 @@
+# Change default value of skip-checks
+- op: replace
+  path: /spec/params/4/default
+  value: true
```
Have you tried running the pipeline? I think this is invalid

ligangty replied:
I've tested the pipeline in my local workspace with corresponding changes and it can pass.

chmeliik replied:
That's weird :thinking: 

```
$ oc kustomize pipelines/maven-zip-build | oc create -f -
Error from server (InternalError): error when creating "STDIN": Internal error occurred: Internal error occurred: conversion webhook for tekton.dev/v1, Kind=Pipeline failed: unable to unmarshal input: json: cannot unmarshal array into Go struct field PipelineSpec.spec.tasks of type v1.PipelineTask
```

(which is probably about something other than this boolean, but it doesn't work)

ligangty replied:
Ok let me have a check.

ligangty replied:
Ok, I think this is cause by a typo in build-maven-zip task adding. An extra dash is added there. I've fixed it in new commit.

chmeliik commented:
I noticed the pipeline has 0 scans, no `snyk`, `clamav`, nothing. Is that OK?

ligangty commented:
> I noticed the pipeline has 0 scans, no `snyk`, `clamav`, nothing. Is that OK?

@chmeliik I confirmed with @ralphbean that all current scan tasks can not work with oci-artifact, that's why I removed all these tasks there. When these tasks have ability to scan the binary content of the oci-artifact we may consider to add them back.

chmeliik commented:
> > I noticed the pipeline has 0 scans, no `snyk`, `clamav`, nothing. Is that OK?
> 
> @chmeliik I confirmed with @ralphbean that all current scan tasks can not work with oci-artifact, that's why I removed all these tasks there. When these tasks have ability to scan the binary content of the oci-artifact we may consider to add them back.

Hmm, I see. We can enable snyk at least, because it doesn't can the oci-artifact, it scans the source repo and the prefetched content

ligangty commented:
> > > I noticed the pipeline has 0 scans, no `snyk`, `clamav`, nothing. Is that OK?
> > 
> > 
> > @chmeliik I confirmed with @ralphbean that all current scan tasks can not work with oci-artifact, that's why I removed all these tasks there. When these tasks have ability to scan the binary content of the oci-artifact we may consider to add them back.
> 
> Hmm, I see. We can enable snyk at least, because it doesn't scan the oci-artifact, it scans the source repo and the prefetched content

Ok, I'll added it and have a try in my workspace.

MartinBasti commented:
> > > I noticed the pipeline has 0 scans, no `snyk`, `clamav`, nothing. Is that OK?
> > 
> > 
> > @chmeliik I confirmed with @ralphbean that all current scan tasks can not work with oci-artifact, that's why I removed all these tasks there. When these tasks have ability to scan the binary content of the oci-artifact we may consider to add them back.
> 
> Hmm, I see. We can enable snyk at least, because it doesn't scan the oci-artifact, it scans the source repo and the prefetched content

Is there a Enterprise Contract policy that will allow maven artifacts to pass without scans?

ligangty commented:
> > > > I noticed the pipeline has 0 scans, no `snyk`, `clamav`, nothing. Is that OK?
> > > 
> > > 
> > > @chmeliik I confirmed with @ralphbean that all current scan tasks can not work with oci-artifact, that's why I removed all these tasks there. When these tasks have ability to scan the binary content of the oci-artifact we may consider to add them back.
> > 
> > 
> > Hmm, I see. We can enable snyk at least, because it doesn't scan the oci-artifact, it scans the source repo and the prefetched content
> 
> Is there a Enterprise Contract policy that will allow maven artifacts to pass without scans?

I think we should create a new one later in release phase to disable the clair-scan.

ligangty commented:
> > > > I noticed the pipeline has 0 scans, no `snyk`, `clamav`, nothing. Is that OK?
> > > 
> > > 
> > > @chmeliik I confirmed with @ralphbean that all current scan tasks can not work with oci-artifact, that's why I removed all these tasks there. When these tasks have ability to scan the binary content of the oci-artifact we may consider to add them back.
> > 
> > 
> > Hmm, I see. We can enable snyk at least, because it doesn't scan the oci-artifact, it scans the source repo and the prefetched content
> 
> Ok, I'll added it and have a try in my workspace.

Added in new commit.

chmeliik commented on pipelines/maven-zip-build-oci-ta/patch.yaml:
```diff
@@ -0,0 +1,94 @@
+    value: $(params.image-expires-after)
+- op: remove
+  path: /spec/tasks/2/workspaces/0
+- op: remove
+  path: /spec/tasks/2/when
```
Why remove the `when` for prefetch-dependencies? The regular build-maven-zip pipeline has it, `-oci-ta` does not

chmeliik replied:
Ah, found 51d7cf9ee96eae89520e8632bf013975dbdbdd65. Makes sense

Long live useful commit messages! :tada:
